### DISCUSSION TITLE
how to do ```COALESCE``` and ```->>``` for json column in Serverpod

### DISCUSSION THREAD

**Discussion by ameerPillsbee** [serverpod/serverpod#1233] [Q&A] [answered]
Imagine I want to make a query like this. order the result based on "created_date_inner" timestamp column inside a json_colum. and if it is null, i want to order based on the column "created_date_main" . if both are non-json columns, i could have used ```orderByList``` . but one column is inside the json colum. how to do it .. do  need raw query?? or any other way?? i want to make it "Serverpodic" 😂 as much as possible 🙆 

```
SELECT *
FROM transportation_details
ORDER BY COALESCE(("json_colum"->>'created_date_inner')::timestamp without time zone, "created_date_main")
```

**Comment by Isakdl** [comment, marked as answer]
Yes you need to do this with a raw query, there is no support to make such a query in dart currently, it would be a nice feature though!

**Reply by ameerPillsbee** [reply]
i have added a custom query builder for that :

```
String buildQuery({
  required String tableName,
  Expression? where,
  int? limit,
  int? offset,
  Column? orderBy,
  List<Order>? orderByList,
  bool orderDescending = false,
  String? coalesceOrderBy,
}) {
  var query = 'SELECT * FROM $tableName';

  if (where != null) {
    query += ' WHERE $where';
  }

  if (coalesceOrderBy != null) {
    query += ' ORDER BY COALESCE($coalesceOrderBy)';
    if (orderDescending) {
      query += ' DESC';
    }
  } else if (orderBy != null) {
    query += ' ORDER BY $orderBy';
    if (orderDescending) {
      query += ' DESC';
    }
  } else if (orderByList != null && orderByList.isNotEmpty) {
    var strList = <String>[];
    for (var order in orderByList) {
      strList.add(order.toString());
    }
    query += ' ORDER BY ${strList.join(',')}';
    if (orderDescending) {
      query += ' DESC';
    }
  }

  if (limit != null) {
    query += ' LIMIT $limit';
  }
  if (offset != null) {
    query += ' OFFSET $offset';
  }
  return query;
}
```


which can be used like :

```
var query = buildQuery(
  tableName: 'transportation_details',
  coalesceOrderBy: '("transportEdit"->>\'transportRequestedAt\')::timestamp without time zone, "transportRequestedAt"',
);
```

for now. but how can i convert all the ```List<List<>>``` back to ```List<TransportationDetail>```  ,,?? is there any ready-made options available for that? i cant find it in docs.. or should i do it manually ??

**Reply by Isakdl** [reply]
I would do a `list.fold()` or `list.expand()` :).
